File: orchestrator/desktop-electron/src/renderer/src/workplace/docflowCreate.ts
import { api } from '../api/client'
import type { UserProfile } from '../api/types'
import { onecGatewayInvokeArgs } from './userContext'

export type DocflowProcessId = 'performance' | 'acquaintance' | 'consideration'

export interface DocflowDocType {
  id: string
  name: string
}

export interface DocflowBasisKind {
  id: string
  label: string
  hint: string
  types: DocflowDocType[]
  processes: DocflowProcessId[]
}

export interface DocflowProcessDef {
  id: DocflowProcessId
  label: string
  hint: string
  multiple: boolean
  verifier: boolean
}

export interface DocflowBasisDocument {
  id: string
  type: string
  name: string
  title: string
  reg_number: string
  reg_date: string
  author: string
  document_type: string
  summary: string
}

export type DocflowPriority = 'high' | 'normal' | 'low'

export const DOCFLOW_PRIORITY_LABEL: Record<DocflowPriority, string> = {
  high: 'Высокий',
  normal: 'Обычный',
  low: 'Низкий'
}

/** Одна строка шага 3 = одна задача (отдельный процесс в ДО). */
export interface DocflowTaskRowDraft {
  key: string
  fio: string
  description: string
  dueDate: string
  dueTime: string
  priority: DocflowPriority
  /** Дата постановки: момент, когда строку добавили. */
  postedAt: Date
}

export interface DocflowLaunchDraft {
  kind: string
  process: DocflowProcessId
  document: DocflowBasisDocument
  title: string
  row: DocflowTaskRowDraft
  verifier: string
  sampleTaskIds: string[]
}

/** +N рабочих дней: суббота и воскресенье не считаются (праздники — нет). */
export function addWorkingDays(from: Date, days: number): Date {
  const result = new Date(from)
  let left = days
  while (left > 0) {
    result.setDate(result.getDate() + 1)
    const weekday = result.getDay()
    if (weekday !== 0 && weekday !== 6) left -= 1
  }
  return result
}

function pad(value: number): string {
  return String(value).padStart(2, '0')
}

export function isoDay(value: Date): string {
  return `${value.getFullYear()}-${pad(value.getMonth() + 1)}-${pad(value.getDate())}`
}

export function clockTime(value: Date): string {
  return `${pad(value.getHours())}:${pad(value.getMinutes())}`
}

export function newTaskRow(now = new Date()): DocflowTaskRowDraft {
  const due = addWorkingDays(now, 2)
  return {
    key: `t-${now.getTime()}-${Math.random().toString(36).slice(2, 7)}`,
    fio: '',
    description: '',
    dueDate: isoDay(due),
    dueTime: clockTime(now),
    priority: 'normal',
    postedAt: now
  }
}

/** Название задачи в ДО: первая строка описания, иначе название из заготовки процесса. */
export function taskTitle(row: DocflowTaskRowDraft, fallback: string): string {
  const first = row.description.trim().split('\n')[0]?.trim() || ''
  return (first || fallback).slice(0, 150)
}

type Result<T> = { ok: true; value: T } | { ok: false; error: string }

async function call<T>(
  user: UserProfile,
  args: Record<string, unknown>,
  timeoutMs: number
): Promise<Result<T>> {
  const res = await api.invokeServerTool('onec.docflow_create', onecGatewayInvokeArgs(user, args), timeoutMs)
  if (!res.ok) return { ok: false, error: res.error || 'Документооборот не ответил' }
  return { ok: true, value: (res.result ?? {}) as T }
}

/** Совпадает с DOC_KINDS / PROCESSES в backend/app/services/docflow_create.py. */
export const DOCFLOW_BASIS_KINDS: DocflowBasisKind[] = [
  { id: 'protocol', label: 'Протокол', hint: 'Решения совещаний и заседаний', types: [], processes: ['performance', 'acquaintance'] },
  {
    id: 'memo',
    label: 'Служебная записка',
    hint: 'Внутренние обращения и запросы',
    types: [],
    processes: ['performance', 'acquaintance', 'consideration']
  },
  { id: 'order', label: 'Приказ', hint: 'Приказы по основной деятельности', types: [], processes: ['acquaintance', 'performance'] },
  { id: 'directive', label: 'Распоряжение', hint: 'Распоряжения руководства', types: [], processes: ['acquaintance', 'performance'] },
  { id: 'contract', label: 'Договор', hint: 'Договоры, доп. соглашения, спецификации', types: [], processes: ['performance'] },
  { id: 'request', label: 'Заявка', hint: 'Заявки и заказы', types: [], processes: ['performance', 'consideration'] },
  { id: 'incoming', label: 'Входящий документ', hint: 'Письма, поступившие в компанию', types: [], processes: ['consideration', 'performance'] },
  { id: 'outgoing', label: 'Исходящий документ', hint: 'Письма, отправленные из компании', types: [], processes: ['performance'] }
]

export const DOCFLOW_PROCESSES: DocflowProcessDef[] = [
  {
    id: 'performance',
    label: 'Исполнение',
    hint: 'Исполнители получат задачу «Исполнить», проверяющий — «Проверить исполнение».',
    multiple: true,
    verifier: true
  },
  { id: 'acquaintance', label: 'Ознакомление', hint: 'Каждый участник получит задачу «Ознакомиться».', multiple: true, verifier: false },
  {
    id: 'consideration',
    label: 'Рассмотрение',
    hint: 'Сотрудник получит «Рассмотреть», вы — «Обработать резолюцию».',
    multiple: false,
    verifier: false
  }
]

let catalogSession: Promise<Result<{ kinds: DocflowBasisKind[]; processes: DocflowProcessDef[] }>> | null = null

/** Точные виды документов из ДО (для уточняющего списка). Один запрос на сеанс. */
export function loadDocflowCatalog(
  user: UserProfile
): Promise<Result<{ kinds: DocflowBasisKind[]; processes: DocflowProcessDef[] }>> {
  if (!catalogSession) {
    catalogSession = call<{ kinds: DocflowBasisKind[]; processes: DocflowProcessDef[] }>(
      user,
      { action: 'catalog' },
      90_000
    ).then((res) => {
      if (!res.ok) catalogSession = null
      return res
    })
  }
  return catalogSession
}

/** Все пользователи ДО (кому можно поставить задачу). */
export async function loadDocflowUsers(user: UserProfile): Promise<Result<{ users: string[] }>> {
  return call(user, { action: 'users' }, 180_000)
}

export async function searchDocflowDocuments(
  user: UserProfile,
  params: {
    kind: string
    documentTypeId: string
    query: string
    onlyMine: boolean
    dateFrom: string
    dateTo: string
  }
): Promise<Result<{ documents: DocflowBasisDocument[] }>> {
  return call(
    user,
    {
      action: 'search_documents',
      kind: params.kind,
      document_type_id: params.documentTypeId,
      query: params.query,
      only_mine: params.onlyMine,
      date_from: params.dateFrom,
      date_to: params.dateTo
    },
    180_000
  )
}

export async function prepareDocflowProcess(
  user: UserProfile,
  params: { kind: string; process: DocflowProcessId; document: DocflowBasisDocument }
): Promise<Result<{ name: string; description: string; author: string; target: string }>> {
  return call(
    user,
    {
      action: 'prepare',
      kind: params.kind,
      process: params.process,
      document: { id: params.document.id, type: params.document.type }
    },
    90_000
  )
}

export async function launchDocflowProcess(
  user: UserProfile,
  draft: DocflowLaunchDraft
): Promise<Result<{ summary: string; performers: string[] }>> {
  const due = `${draft.row.dueDate}T${draft.row.dueTime || '18:00'}`
  return call(
    user,
    {
      action: 'launch',
      confirm: true,
      kind: draft.kind,
      process: draft.process,
      document: { id: draft.document.id, type: draft.document.type },
      title: draft.title.trim(),
      description: draft.row.description.trim(),
      due,
      priority: draft.row.priority,
      verifier: draft.verifier.trim(),
      sample_task_ids: draft.sampleTaskIds,
      performers: [{ fio: draft.row.fio.trim(), due, note: '' }]
    },
    180_000
  )
}

export function formatRuDate(value: string): string {
  const match = /^(\d{4})-(\d{2})-(\d{2})/.exec(value || '')
  return match ? `${match[3]}.${match[2]}.${match[1]}` : value || '—'
}

export function formatRuDateTime(value: Date): string {
  return `${formatRuDate(isoDay(value))} ${clockTime(value)}`
}

export function documentLabel(doc: DocflowBasisDocument): string {
  return doc.name || doc.title || doc.reg_number || 'Документ'
}

export interface DocflowApprovalItem {
  name: string
  position: string
  date: string
  result: string
  comment: string
}

export interface DocflowApprovalSheet {
  found: boolean
  documentName: string
  regNumber: string
  items: DocflowApprovalItem[]
  message: string
}

const approvalCache = new Map<string, DocflowApprovalSheet>()
const approvalLoads = new Map<string, Promise<DocflowApprovalSheet>>()
let approvalBlocked = ''

function emptySheet(message: string): DocflowApprovalSheet {
  return { found: false, documentName: '', regNumber: '', items: [], message }
}

/** Лист согласования документа ДО. Один запрос на документ за сеанс. */
export function loadDocflowApprovalSheet(
  user: UserProfile,
  params: { id: string; kind: 'order' | 'directive'; regNumber: string; title: string }
): Promise<DocflowApprovalSheet> {
  if (approvalBlocked) return Promise.resolve(emptySheet(approvalBlocked))
  const cached = approvalCache.get(params.id)
  if (cached) return Promise.resolve(cached)
  const pending = approvalLoads.get(params.id)
  if (pending) return pending
  const load = call<{
    found?: boolean
    document_name?: string
    reg_number?: string
    items?: DocflowApprovalItem[]
  }>(
    user,
    {
      action: 'approval_sheet',
      kind: params.kind,
      reg_number: params.regNumber,
      title: params.title
    },
    120_000
  )
    .then((res) => {
      approvalLoads.delete(params.id)
      if (!res.ok) {
        if (/парол|учётн|учетн|401|403/i.test(res.error)) approvalBlocked = res.error
        const sheet = emptySheet(res.error)
        approvalCache.set(params.id, sheet)
        return sheet
      }
      const value = res.value
      const sheet: DocflowApprovalSheet = {
        found: value.found === true,
        documentName: value.document_name || '',
        regNumber: value.reg_number || '',
        items: Array.isArray(value.items) ? value.items : [],
        message: value.found === true ? '' : 'В документообороте нет листа согласования по этому номеру'
      }
      approvalCache.set(params.id, sheet)
      return sheet
    })
    .catch((err: unknown) => {
      approvalLoads.delete(params.id)
      const sheet = emptySheet(err instanceof Error ? err.message : 'Лист согласования не загрузился')
      approvalCache.set(params.id, sheet)
      return sheet
    })
  approvalLoads.set(params.id, load)
  return load
}

File: backend/app/services/docflow_create.py
"""Создание задачи в 1С:Документооборот: документ-основание → процесс → запуск."""

from __future__ import annotations

import logging
import re
from typing import Any

from app.services.docflow_task_action import _session_credentials
from app.services.docflow_tasks import DocflowError
from app.tools.onec.dok_create import (
    fill_process,
    launch_business_process,
    list_internal_document_types,
    list_users,
    new_business_process,
    open_task_ids_by_step,
    parse_due,
    sample_performer,
    search_documents,
)
from app.tools.onec.dok_soap import find_user, load_config, xml_text

logger = logging.getLogger(__name__)

DOC_KINDS: list[dict[str, Any]] = [
    {
        "id": "protocol",
        "label": "Протокол",
        "hint": "Решения совещаний и заседаний",
        "dm_type": "DMInternalDocument",
        "pattern": r"протокол",
        "processes": ["performance", "acquaintance"],
    },
    {
        "id": "memo",
        "label": "Служебная записка",
        "hint": "Внутренние обращения и запросы",
        "dm_type": "DMInternalDocument",
        "pattern": r"служебн",
        "processes": ["performance", "acquaintance", "consideration"],
    },
    {
        "id": "order",
        "label": "Приказ",
        "hint": "Приказы по основной деятельности",
        "dm_type": "DMInternalDocument",
        "pattern": r"приказ",
        "processes": ["acquaintance", "performance"],
    },
    {
        "id": "directive",
        "label": "Распоряжение",
        "hint": "Распоряжения руководства",
        "dm_type": "DMInternalDocument",
        "pattern": r"распоряж",
        "processes": ["acquaintance", "performance"],
    },
    {
        "id": "contract",
        "label": "Договор",
        "hint": "Договоры, доп. соглашения, спецификации",
        "dm_type": "DMInternalDocument",
        "pattern": r"договор|соглашени|спецификац",
        "processes": ["performance"],
    },
    {
        "id": "request",
        "label": "Заявка",
        "hint": "Заявки и заказы",
        "dm_type": "DMInternalDocument",
        "pattern": r"заявк|заказ",
        "processes": ["performance", "consideration"],
    },
    {
        "id": "incoming",
        "label": "Входящий документ",
        "hint": "Письма, поступившие в компанию",
        "dm_type": "DMIncomingDocument",
        "author_field": "responsible",
        "processes": ["consideration", "performance"],
    },
    {
        "id": "outgoing",
        "label": "Исходящий документ",
        "hint": "Письма, отправленные из компании",
        "dm_type": "DMOutgoingDocument",
        "processes": ["performance"],
    },
]

PROCESSES: dict[str, dict[str, Any]] = {
    "performance": {
        "label": "Исполнение",
        "dm_type": "DMBusinessProcessPerformance",
        "hint": "Исполнители получат задачу «Исполнить», проверяющий — «Проверить исполнение».",
        "multiple": True,
        "verifier": True,
    },
    "acquaintance": {
        "label": "Ознакомление",
        "dm_type": "DMBusinessProcessAcquaintance",
        "hint": "Каждый участник получит задачу «Ознакомиться».",
        "multiple": True,
        "verifier": False,
    },
    "consideration": {
        "label": "Рассмотрение",
        "dm_type": "DMBusinessProcessConsideration",
        "hint": "Сотрудник получит «Рассмотреть», вы — «Обработать резолюцию».",
        "multiple": False,
        "verifier": False,
    },
}

_LOOKUP_TIMEOUT = 60.0
_SEARCH_TIMEOUT = 120.0


def _kind(kind_id: str) -> dict[str, Any]:
    for item in DOC_KINDS:
        if item["id"] == kind_id:
            return item
    raise DocflowError(f"Неизвестный вид документа-основания: {kind_id or '—'}")


def _process(kind: dict[str, Any], process_id: str) -> dict[str, Any]:
    if process_id not in PROCESSES:
        raise DocflowError(f"Неизвестный процесс: {process_id or '—'}")
    if process_id not in kind["processes"]:
        raise DocflowError(f"Для «{kind['label']}» процесс «{PROCESSES[process_id]['label']}» не используется")
    return PROCESSES[process_id]


def _config(args: dict[str, Any], actor_fio: str):
    user, password = _session_credentials(args)
    if not password:
        user = user or actor_fio.strip()
    if not user or not password:
        raise DocflowError("Нужны учётные данные сеанса 1С (ФИО и пароль)")
    return user, load_config(username=user, password=password, require_user=True)


def _resolve_user(config, fio: str) -> dict[str, str]:
    try:
        return find_user(config, fio)
    except ValueError as exc:
        raise DocflowError(str(exc)) from exc


def _types_for(kind: dict[str, Any], types: list[dict[str, str]]) -> list[dict[str, str]]:
    pattern = kind.get("pattern")
    if not pattern:
        return []
    return [item for item in types if re.search(pattern, item["name"], flags=re.I)]


def _catalog(config) -> dict[str, Any]:
    try:
        types = list_internal_document_types(config, timeout=_LOOKUP_TIMEOUT)
    except RuntimeError as exc:
        raise DocflowError(f"Виды документов ДО не получены: {exc}") from exc
    kinds = []
    for kind in DOC_KINDS:
        matched = _types_for(kind, types)
        if kind["dm_type"] == "DMInternalDocument" and not matched:
            continue
        kinds.append(
            {
                "id": kind["id"],
                "label": kind["label"],
                "hint": kind["hint"],
                "types": matched,
                "processes": kind["processes"],
            }
        )
    processes = [
        {"id": pid, "label": p["label"], "hint": p["hint"], "multiple": p["multiple"], "verifier": p["verifier"]}
        for pid, p in PROCESSES.items()
    ]
    return {"summary": "Каталог оснований ДО", "kinds": kinds, "processes": processes}


def _search(config, args: dict[str, Any], actor: str) -> dict[str, Any]:
    kind = _kind(str(args.get("kind") or ""))
    only_mine = args.get("only_mine") is not False
    author = _resolve_user(config, actor) if only_mine else None
    type_ids = [str(args.get("document_type_id") or "").strip()]
    if kind["dm_type"] == "DMInternalDocument" and not type_ids[0]:
        types = _types_for(kind, list_internal_document_types(config, timeout=_LOOKUP_TIMEOUT))
        type_ids = [item["id"] for item in types[:6]]
        if not type_ids:
            raise DocflowError(f"В ДО нет видов документов для «{kind['label']}»")
    rows: list[dict[str, str]] = []
    seen: set[str] = set()
    for type_id in type_ids:
        try:
            found = search_documents(
                config,
                kind["dm_type"],
                document_type_id=type_id,
                author=author,
                author_field=kind.get("author_field", "author"),
                query=str(args.get("query") or ""),
                limit=int(args.get("limit") or 50),
                timeout=_SEARCH_TIMEOUT,
            )
        except RuntimeError as exc:
            raise DocflowError(f"Поиск документов не выполнен: {exc}") from exc
        for row in found:
            if row["id"] not in seen:
                seen.add(row["id"])
                rows.append(row)
    rows.sort(key=lambda row: row.get("reg_date") or "", reverse=True)
    return {"summary": f"Найдено документов: {len(rows)}", "documents": rows, "count": len(rows)}


def _target(args: dict[str, Any]) -> dict[str, str]:
    doc = args.get("document") if isinstance(args.get("document"), dict) else {}
    target = {"id": str(doc.get("id") or "").strip(), "type": str(doc.get("type") or "").strip()}
    if not target["id"] or not target["type"]:
        raise DocflowError("Не выбран документ-основание")
    return target


def _prepare(config, args: dict[str, Any]) -> dict[str, Any]:
    kind = _kind(str(args.get("kind") or ""))
    process = _process(kind, str(args.get("process") or ""))
    try:
        node = new_business_process(config, process["dm_type"], _target(args), timeout=_LOOKUP_TIMEOUT)
    except RuntimeError as exc:
        raise DocflowError(f"ДО не подготовил процесс: {exc}") from exc
    return {
        "summary": "Заготовка процесса получена",
        "name": xml_text(node, "m:name"),
        "description": xml_text(node, "m:description"),
        "author": xml_text(node, "m:author/m:name"),
        "target": xml_text(node, "m:target/m:name"),
    }


def _launch(config, args: dict[str, Any], actor: str) -> dict[str, Any]:
    if args.get("confirm") is not True:
        raise DocflowError("Запуск процесса требует подтверждения")
    kind = _kind(str(args.get("kind") or ""))
    process = _process(kind, str(args.get("process") or ""))
    target = _target(args)
    title = str(args.get("title") or "").strip()
    description = str(args.get("description") or "").strip()
    if not title:
        raise DocflowError("Укажите название задачи")
    try:
        due = parse_due(args.get("due"))
    except ValueError as exc:
        raise DocflowError(str(exc)) from exc
    priority = str(args.get("priority") or "").strip().lower()
    if priority and priority not in {"high", "normal", "low"}:
        raise DocflowError("Приоритет: high | normal | low")

    raw_performers = args.get("performers") if isinstance(args.get("performers"), list) else []
    performers: list[dict[str, Any]] = []
    for raw in raw_performers:
        if not isinstance(raw, dict) or not str(raw.get("fio") or "").strip():
            continue
        try:
            personal_due = parse_due(raw.get("due"))
        except ValueError as exc:
            raise DocflowError(str(exc)) from exc
        performers.append(
            {
                "user": _resolve_user(config, str(raw["fio"]).strip()),
                "due": personal_due,
                "note": str(raw.get("note") or "").strip(),
            }
        )
    if not performers:
        raise DocflowError("Укажите хотя бы одного исполнителя")
    if not process["multiple"] and len(performers) > 1:
        raise DocflowError(f"«{process['label']}» назначается одному сотруднику")

    verifier = None
    if process["verifier"]:
        verifier = _resolve_user(config, str(args.get("verifier") or actor).strip())

    template = None
    if process["dm_type"] == "DMBusinessProcessPerformance":
        ids = [str(item) for item in (args.get("sample_task_ids") or []) if str(item).strip()]
        ids += [item for item in open_task_ids_by_step(r"^исполнить$") if item not in ids]
        template = sample_performer(config, process["dm_type"], ids, timeout=_LOOKUP_TIMEOUT)
        logger.info("dok_create: образец участника %s", "найден" if template is not None else "не найден")

    try:
        node = new_business_process(config, process["dm_type"], target, timeout=_LOOKUP_TIMEOUT)
        fill_process(
            node,
            process["dm_type"],
            title=title,
            description=description,
            due=due,
            performers=performers,
            verifier=verifier,
            performer_template=template,
            priority=priority,
        )
        launched = launch_business_process(config, node, timeout=_LOOKUP_TIMEOUT)
    except ValueError as exc:
        raise DocflowError(str(exc)) from exc
    except RuntimeError as exc:
        raise DocflowError(f"ДО не запустил процесс: {exc}") from exc

    names = ", ".join(item["user"]["name"] for item in performers)
    return {
        "summary": f"Процесс «{process['label']}» запущен в документообороте. Задачи получат: {names}.",
        "process": launched,
        "performers": [item["user"]["name"] for item in performers],
    }


def handle_docflow_create(
    args: dict[str, Any],
    *,
    actor_fio: str = "",
    actor_user_id: str = "",
) -> dict[str, Any]:
    del actor_user_id
    action = str(args.get("action") or "").strip().lower()
    actor, config = _config(args, actor_fio)
    if action == "catalog":
        return _catalog(config)
    if action == "users":
        try:
            names = list_users(config, timeout=_SEARCH_TIMEOUT)
        except RuntimeError as exc:
            raise DocflowError(f"Пользователи ДО не получены: {exc}") from exc
        return {"summary": f"Пользователей ДО: {len(names)}", "users": names, "count": len(names)}
    if action == "search_documents":
        return _search(config, args, actor)
    if action == "prepare":
        return _prepare(config, args)
    if action == "launch":
        return _launch(config, args, actor)
    raise DocflowError("Укажите action: catalog | users | search_documents | prepare | launch")


def stub_docflow_create(args: dict[str, Any], **_: Any) -> dict[str, Any]:
    return {
        "summary": "stub: docflow_create",
        "action": str(args.get("action") or ""),
        "ok": False,
        "message": "Настройте DOK_HTTP_* на backend для создания задач в документообороте",
    }
